### ISSUE TITLE
reduce number of times rendered attribute value expression is evaluated

### ISSUE THREAD

**Issue by glassfishrobot** [javaee/javaserverfaces-spec#941]
The rendered attribute is an Achilles heal of JSF performance. JSF abuses the
value expression used in rendered attributes. It's expected that the value
expression is going to be resolved more than once on a request because by nature
it is a pass-through construct and because the value it is representing is
likely contextual (meaning it can change as the application changes). However,
in the case of the render attribute, it's a little bit out of control. Any
getter method bound to a rendered attribute is being called way too many times.
This could easily be solved by being more frugal about when it is consulted.

For example, the rendered value expression is resolved in encodeChildren,
encodeBegin (or encodeParentAndChildren) and again at encodeEnd of most
components. We should state that when a component is first addressed during
encoding (perhaps in encodeBegin) that is when the rendered attribute should be
evaluated. Then the resolved value simply cannot change again in that phase.
(The exception would be a row in UIData, which the rendered attribute would need
to be resolved once per component per row).

We should also consider having encodeAll check the rendered attribute and not
encodeBegin. Or, we could keep the current contract the same but deprecate
encodeBegin, encodeChildren, encodeEnd and getRendersChildren so that most
people start using encodeAll instead so that isRendered gets called only once
per component per render phase execution.
#### Affected Versions
[2.2]

**Comment by glassfishrobot** [comment]
Reported by rogerk

**Comment by glassfishrobot** [comment]
@edburns said:
Set fixVersion to 2.3\. I think this may be better handled by having a way to advise the EL to start caching expression evaluations for the current thread and have a way to turn off that caching at the end of the JSF lifecycle.

**Comment by glassfishrobot** [comment]
@edburns said:
Set priority to baseline ahead of JSF 2.3 triage. Priorities will be assigned accurately after this exercise.

**Comment by glassfishrobot** [comment]
@manfredriem said:
Setting priority to Critical

**Comment by glassfishrobot** [comment]
@manfredriem said:
As per EG discussion I am closing this as "Won't fix".

**Comment by glassfishrobot** [comment]
Marked as **won't fix** on Monday, January 26th 2015, 8:10:19 am

**Comment by glassfishrobot** [comment]
@BalusC said:
For sake of completeness, here's the EG discussion link: [https://java.net/projects/javaserverfaces-spec-public/lists/jsr372-experts/archive/2015-01/message/79](https://java.net/projects/javaserverfaces-spec-public/lists/jsr372-experts/archive/2015-01/message/79)

**Comment by glassfishrobot** [comment]
This issue was imported from java.net JIRA JAVASERVERFACES_SPEC_PUBLIC-941
